File: compress/src/java/compresslib/Mtf.java
package compresslib;

/**
 * move-to-front — SPEC §4.
 *
 * <p>앞으로 <b>옮기는</b> 것이지 바꿔치는 것이 아니다. 바꿔치기도
 * 자기들끼리는 왕복이 되므로, 골든 벡터가 없으면 갈라진 줄도 모른다.
 */
public final class Mtf {
  public static final int ALPHABET = 256;

  private Mtf() {}

  private static byte[] identity() {
    byte[] t = new byte[ALPHABET];
    for (int i = 0; i < ALPHABET; i++) {
      t[i] = (byte) i;
    }
    return t;
  }

  public static byte[] transform(byte[] src) {
    byte[] table = identity();
    byte[] out = new byte[src.length];
    for (int p = 0; p < src.length; p++) {
      byte b = src[p];
      int i = 0;
      while (table[i] != b) {
        i++;
      }
      out[p] = (byte) i;
      for (int k = i; k > 0; k--) {
        table[k] = table[k - 1];
      }
      table[0] = b;
    }
    return out;
  }

  public static byte[] inverse(byte[] src) {
    byte[] table = identity();
    byte[] out = new byte[src.length];
    for (int p = 0; p < src.length; p++) {
      int idx = src[p] & 0xFF;
      byte b = table[idx];
      out[p] = b;
      for (int k = idx; k > 0; k--) {
        table[k] = table[k - 1];
      }
      table[0] = b;
    }
    return out;
  }

  public static byte[] encode(byte[] src) {
    ByteBuf out = new ByteBuf();
    Varint.put(out, src.length);
    out.extend(transform(src));
    return out.bytes();
  }

  public static byte[] decode(byte[] src) {
    int[] h = Varint.getLength(src, 0);
    if (src.length - h[1] != h[0]) {
      throw CodecException.of("몸통 길이가 헤더와 다르다: %d != %d",
          src.length - h[1], h[0]);
    }
    byte[] body = java.util.Arrays.copyOfRange(src, h[1], src.length);
    return inverse(body);
  }
}

File: compress/src/py/compresslib/mtf.py
# -*- coding: utf-8 -*-
"""move-to-front — SPEC §4.

방금 본 바이트를 표의 맨 앞으로 옮긴다. 그러면 "최근에 본 것" 이 작은
숫자가 되고, 같은 바이트가 몰려 있는 열은 0 이 줄줄이 나오는 열이 된다.
BWT 뒤에 붙이면 그 성질이 극대화된다 — 10부의 본론이 그것이다.

이것만으로는 한 바이트도 안 준다. 길이가 그대로인 **치환**이기 때문이다.
줄이는 일은 그 다음에 오는 0런 부호와 허프만이 한다.

옮기는 것이지 바꿔치는 것이 아니다. 바꿔치기(swap)도 자기들끼리는 왕복이
되기 때문에, 골든 벡터가 없으면 다섯 언어가 갈라진 줄도 모른다.

표를 그냥 리스트로 둔다 — 최악 O(n·256). 이 덱이 다루는 크기에서는
충분하고, 무엇보다 **덱에 실린 코드가 곧 숫자를 낸 코드** 여야 한다.
더 빠른 변형(자기 균형 트리·비트 벡터)은 10부에서 말로만 다룬다.
"""
from compresslib import varint

ALPHABET = 256


def transform(src):
    table = list(range(ALPHABET))
    out = bytearray()
    for b in src:
        i = table.index(b)
        out.append(i)
        if i:
            del table[i]
            table.insert(0, b)
    return bytes(out)


def inverse(src):
    table = list(range(ALPHABET))
    out = bytearray()
    for i in src:
        b = table[i]
        out.append(b)
        if i:
            del table[i]
            table.insert(0, b)
    return bytes(out)


def encode(src):
    return varint.put(len(src)) + transform(src)


def decode(src):
    n, pos = varint.get_length(src)
    body = src[pos:]
    if len(body) != n:
        raise ValueError('몸통 길이가 헤더와 다르다: %d != %d'
                         % (len(body), n))
    return inverse(body)
